test/object-store/sync/flx_role_change.cpp

```diff
@@ -0,0 +1,861 @@
+            // Make sure a client reset did not occur while waiting for the role change to
+            // be applied
+            FAIL("Client reset is not expected when the role/rules/permissions are changed");
+        };
+    };
+
+    auto update_perms_and_verify = [&](FLXSyncTestHarness& harness, SharedRealm check_realm, nlohmann::json new_rules,
+                                       ExpectedResults expected) {
+        // Reset the state machine
+        state_machina.transition_with([&](TestState cur_state) {
+            REQUIRE(cur_state == TestState::not_ready);
+            bootstrap_msg_count = 0;
+            download_msg_count = 0;
+            role_change_bootstrap = false;
+            query_version = check_realm->get_active_subscription_set().version();
+            if (expected.bootstrap == BootstrapMode::SingleMessageMulti) {
+                send_test_command = true;
+            }
+            return TestState::start;
+        });
+
+        // Update the permissions on the server - should send an error to the client to force
+        // it to reconnect
+        auto& app_session = harness.session().app_session();
+        logger->debug("ROLE CHANGE: Updating rule definitions: %1", new_rules);
+        app_session.admin_api.update_default_rule(app_session.server_app_id, new_rules);
+
+        if (expected.bootstrap != BootstrapMode::NoReconnect) {
+            // After updating the permissions (if they are different), the server should send an
+            // error that will disconnect/reconnect the session - verify the reconnect occurs.
+            // Make sure at least the reconnect state (or later) has been reached
+            auto state_reached = state_machina.wait_until([](TestState cur_state) {
+                return static_cast<int>(cur_state) >= static_cast<int>(TestState::reconnect_received);
+            });
+            REQUIRE(state_reached);
+        }
+
+        // Assuming the session disconnects and reconnects, the server initiated role change
+        // bootstrap download will take place when the session is re-established and will
+        // complete before the server sends the initial MARK response.
+        REQUIRE(!wait_for_download(*check_realm));
+        REQUIRE(!wait_for_upload(*check_realm));
+
+        // Now that the server initiated bootstrap should be complete, verify the operation
+        // performed matched what was expected.
+        state_machina.transition_with([&](TestState cur_state) {
+            switch (expected.bootstrap) {
+                case BootstrapMode::NoReconnect:
+                    // Confirm that the session did receive an error and a bootstrap did not occur
+                    REQUIRE(cur_state == TestState::start);
+                    REQUIRE_FALSE(role_change_bootstrap);
+                    break;
+                case BootstrapMode::None:
+                    // Confirm that a bootstrap nor a client reset did not occur
```

Comment:
So this is something like NoBootstrapReceivedError? "Confirm that neither a bootstrap nor a client rest did not occur" seems like a double negative?

Reply:
Renamed state to `GotErrorNoBootstrap` and updated comment